### DISCUSSION TITLE
Is 'driver ahead' information accurate?

### DISCUSSION THREAD

**Discussion by thumplik** [theOehrly/Fast-F1#257] [Q&A]
I was looking at Alonso's race in Zandvoort, and I noticed that the driver ahead information doesn't seem to be in sync with what happened in the face (for example, it doesn't show him behind Perez at the point that he should be, it shows him behind Lewis or Sainz). Am I interpreting this field incorrectly?

**Comment by theOehrly** [comment]
There are some issues that I know of with the implementation. They are mostly related to cars in the pitlane and cars on the mainstraight.

Apart from that, to clarify, the position is given to the car physically in front, that means potentially to a lapped car. Is this maybe the problem?

**Reply by thumplik** [reply]
I don't think that's the issue. I think there may be a problem with the distance field. It seems like it isn't an accurate calculation of the distance a driver has driven. Is it a calculated field or reported by telemetry?

**Reply by thumplik** [reply]
For example, walk through De Vries race in Monza. He rarely (if ever) is in his corresponding race position. Some drivers seem to be correct, while others are not and I cannot figure out why.

**Reply by theOehrly** [reply]
Distance to driver ahead is calculated by FastF1. The implementation is not perfect, but the calculation is not quite trivial.

How are you accessing/creating the distance to driver ahead data? Can you preferably give me a short snippet of code? There's a big difference if you access it for individual laps or for many laps at once. (Individual laps are preferred!)

**Reply by thumplik** [reply]
The distance calculation is far from trivial! I appreciate all the work you put into it. I have some ideas on how to calculate it, but I'd rather not... hehe.

So I am trying to monitor each driver over the course of a race. The follow is the code to get the dataframe of all race laps a driver completes in a given race. A bit inelegant, yes. Are there better ways of getting this info?

#monza race
year = 2022
wknd = 16
ses = 'Race'
session = ff1.get_session(year, wknd, ses)
weekend = session.event
session.load()

i=1
for car in ['1', '16', '63', '55', '44', '11', '4', '10', '45', '24', '31', '47', '77', '22', '6', '20', '3', '18', '14', '5']:
    if i<1:
        df = session.laps.pick_driver(car).telemetry
        df['Driver'] = car #for tracking which driver the laps are associated with
        i+=1
    else:
        df_2 = session.laps.pick_driver(car).telemetry
        df_2['Driver'] = car #for tracking which driver the laps are associated with
        df = pd.concat([df, df_2])
